src/Wait.hs

```diff
@@ -53,57 +56,65 @@ listContentsInside test dir = do
         case waiter of
             WaiterPoll t -> return ()
             WaiterNotify manager kick mp -> do
-                dirs <- fmap Set.fromList $ mapM canonicalizePathSafe $ nubOrd $ map takeDirectory files
+                dirs <-
+                  fmap Set.fromList $ mapM (firstM canonicalizePathSafe) $ nubOrd $ map (first takeDirectory) files
                 modifyVar_ mp $ \mp -> do
                     let (keep,del) = Map.partitionWithKey (\k v -> k `Set.member` dirs) mp
                     sequence_ $ Map.elems del
-                    new <- forM (Set.toList $ dirs `Set.difference` Map.keysSet keep) $ \dir -> do
+                    new <- forM (Set.toList $ dirs `Set.difference` Map.keysSet keep) $ \(dir, a) -> do
                         can <- watchDir manager (fromString dir) (const True) $ \event -> do
                             whenLoud $ outStrLn $ "%NOTIFY: " ++ show event
                             void $ tryPutMVar kick ()
-                        return (dir, can)
+                        return ((dir, a), can)
                     let mp2 = keep `Map.union` Map.fromList new
-                    whenLoud $ outStrLn $ "%WAITING: " ++ unwords (Map.keys mp2)
+                    whenLoud $ outStrLn $ "%WAITING: " ++ unwords (map fst (Map.keys mp2))
                     return mp2
                 void $ tryTakeMVar kick
-        new <- mapM getModTime files
+        new <- mapM (getModTime . fst) files
         case [x | (x,Just t) <- zip files new, t > base] of
-            [] -> recheck files new
-            xs -> return xs
-    where
-        recheck files old = do
+            [] -> Right <$> recheck files new
+            xs -> return (Right xs)
+
+    recheck :: [(FilePath, a)] -> [Maybe UTCTime] -> IO [(String, a)]
+    recheck files old = do
             sleep 0.1
             case waiter of
                 WaiterPoll t -> sleep $ max 0 $ t - 0.1 -- subtract the initial 0.1 sleep from above
                 WaiterNotify _ kick _ -> do
                     takeMVar kick
                     whenLoud $ outStrLn "%WAITING: Notify signaled"
-            new <- mapM getModTime files
+            new <- mapM (getModTime . fst) files
             case [x | (x,t1,t2) <- zip3 files old new, t1 /= t2] of
                 [] -> recheck files new
                 xs -> do
                     let disappeared = [x | (x, Just _, Nothing) <- zip3 files old new]
-                    when (disappeared /= []) $ do
+                    unless (null disappeared) $ do
                         -- if someone is deleting a needed file, give them some space to put the file back
                         -- typically caused by VIM
                         -- but try not to
-                        whenLoud $ outStrLn $ "%WAITING: Waiting max of 1s due to file removal, " ++ unwords disappeared
+                        whenLoud $ outStrLn $ "%WAITING: Waiting max of 1s due to file removal, " ++ unwords (map fst disappeared)
```

Comment:
Does this need a `nubOrd`, since if the file is being waited for twice, as Reload and Restart, it might appear twice in the output?

Reply:
Good point! Fixed!